Full functionality for sim_ancestry

jeromekelleher opened the pull request:
Closes #1171

This is a complete-ish proposal for the new ``sim_ancestry`` function - see below for summary. I've requested reviews from a whole bunch of people. Don't worry if you don't have time to go through everything (or at all), I just wanted to keep people in the loop and given them an early chance to complain if they hate the new interface!

jeromekelleher commented:
This is ready for review now, I think. It needs documentation, obviously, but I'm hoping the code is fairly clear about what all the various corner cases etc are and how we're trying to accomodate them. I'll file some follow-up issues to track the stuff that's come up, and for documenting the API.

The main differences from simulate are:

- Samples refer to k-ploid *individuals* and we fill out the individual table
- Renamed a few parameters so they are clearer, ``length`` -> ``sequence_length``, ``Ne``->``population_size``
- Also renamed ``from_ts`` to ``initial_state``, as this is much more descriptive of what it actually *is* now. I'm imagining documenting this as an advanced way of initialising your simulations, as well as recapitating.
- Much more use of type polymorphism on the input values, so we now just have ``recombination_rate`` which can be either a single value or a RateMap instance.
- ``discrete_genome`` is True by default.

I tried to refactor things a bit so that we're not duplicating too much code between ``sim_ancestry`` and ``simulate``.

petrelharp commented on msprime/ancestry.py:
```diff
@@ -721,113 +762,375 @@ def simulate(
+    node_individual = np.zeros(N, dtype=np.int32)
+    a = np.arange(n, dtype=node_individual.dtype)
+    for j in range(ploidy):
+        node_individual[j::ploidy] = a
```
shorter (and arguably easier to read?) is
```
node_individual = np.repeat(range(n), ploidy)
```

jeromekelleher replied:
Aha, yes, much better, thanks.

petrelharp commented on msprime/ancestry.py:
```diff
@@ -721,113 +762,375 @@ def simulate(
+    # set it.
+    # TODO Might also be worth considering setting ploidy=2 by default. This
+    # *is* likely to be the most common application, and those doing haploid
+    # stuff can set ploidy=1, if they like.
```
I support a default of ploidy=2.

JereKoskela replied:
I agree. It also makes the defaults consistent with ms, which might be helpful for users.

jeromekelleher replied:
Done. Thanks for the input.

petrelharp commented on msprime/ancestry.py:
```diff
@@ -721,113 +762,375 @@ def simulate(
+        random_generator=random_generator,
+        recombination_map=recombination_map,
+        gene_conversion_map=gene_conversion_map,
+        gene_conversion_track_length=gene_conversion_track_length,
```
I think it's "tract" not "track": [891 hits](https://scholar.google.com/scholar?q="gene+conversion+tract") versus [22](https://scholar.google.com/scholar?q="gene+conversion+track").

jeromekelleher replied:
Opened issue #1202 to track (or tract?)

petrelharp commented on msprime/demography.py:
```diff
@@ -225,6 +236,17 @@ def from_old_style(
+        Returns a simple single-population model.
+        """
+        check_population_size(population_size)
+        # TODO what should the population name be here??
```
`pop0` seems good?

petrelharp reviewed (approved):
Looks great! I've looked over everything, and it seems to do what we want, except the details about populations that aren't solid yet. I'm a bit worried that so many existing tests are switched to using `_parse_sim_ancestry` instead of `sim_ancestry`, though?

petrelharp commented on msprime/ancestry.py:
```diff
@@ -721,113 +762,375 @@ def simulate(
+        # The most natural way to do this would be a numpy array of individual
+        # IDs. By checking the iterable condition first, we make sure that we're
+        # not interpreting input numpy arrays of length 1 as a numeric argument.
+        # Interpret as a list of Sample objects.
```
maybe the following should be done with numpy, also, e.g.
```
pops = np.array([s.population for s in samples])
node_pops = np.repeat(pops, ploidy)
node_inds = np.repeat(range(n), ploidy)
... set_columns...
```

jeromekelleher replied:
Good call. Opened #1211 to track.

JereKoskela commented on msprime/ancestry.py:
```diff
@@ -721,113 +762,375 @@ def simulate(
+    else:
+        raise TypeError("demography argument must be an instance of msprime.Demography")
+    # TODO should we validate that the population sizes are positive integers
+    # for DTWF? It's not actually required, but could help catch some errors?
```
Is there a reason to not do this for an explicitly finite population model?

jeromekelleher replied:
Mainly because it gets caught at runtime during the simulation - we'll get a "population size has gone to zero" (or similar) error. It's slightly murky, because DTWF rounds the population sizes because of exponential growth, so we do actually handle non-integer sizes within the simulation. It seems a bit odd to say you must have integer sizes on the way in, then, and could end up being awkward. :shrug: I don't know though.

JereKoskela replied:
Good point about needing to round anyway. In that case, I think it feels cleaner to not bother with a check here either.

JereKoskela commented:
Most of these changes are to parts of code I've never looked at, but the `sim_ancestry` interface seems very clear, and definitely covers everything we might want for multiple merger models.

jeromekelleher commented:
>  I'm a bit worried that so many existing tests are switched to using _parse_sim_ancestry instead of sim_ancestry, though?

These are all new tests @petrelharp - all the old tests are untouched. Since the actually simulations are already pretty well tested, I focused on validating the argument parsing side of things.

grahamgower commented on msprime/ancestry.py:
```diff
@@ -721,113 +762,375 @@ def simulate(
+    # confusing problems for (misguided) users of truthiness.
+    discrete_genome = True if discrete_genome is None else bool(discrete_genome)
+    record_migrations = False if record_migrations is None else bool(record_migrations)
+    record_full_arg = False if record_full_arg is None else bool(record_full_arg)
```
Could use a coercion function like this?
```python
def coerce_tristate(b, meaning_of_None):
    if b is None:
        return meaning_of_None
    if not isinstance(b, bool):
        raise TypeError("expected True, False, or None")
    return b

# truth table
assert(True == coerce_tristate(None, True))
assert(True == coerce_tristate(True, True))
assert(False == coerce_tristate(False, True))
assert(False == coerce_tristate(None, False))
assert(True == coerce_tristate(True, False))
assert(False == coerce_tristate(False, False))
```

jeromekelleher replied:
As in, enforce lack of truthiness, so people have to cast to bool explicitly themselves if they want it? Good idea...

jeromekelleher replied:
Done, ta.

grahamgower commented on msprime/ancestry.py:
```diff
@@ -721,113 +762,375 @@ def simulate(
+            raise ValueError("Cannot specify both samples and initial_state")
+        if sequence_length != initial_state.sequence_length:
+            raise ValueError(
+                "The initial_state sequence length must agree with other parameters"
```
It might help the user to see a list of other parameters that this must agree with. Although I suppose such a list might get longer in the future (e.g if a mutation map is added), so this risks becoming outdated and thus confusing if the problem parameter isn't in the list.

grahamgower commented on msprime/ancestry.py:
```diff
@@ -721,113 +762,375 @@ def simulate(
+            )
+        if len(initial_state.populations) == 0:
+            raise ValueError(
+                "initial_state tables must defined at least one population."
```
s/defined/define/

grahamgower reviewed (approved):
LGTM!

gtsambos commented on msprime/ancestry.py:
```diff
@@ -721,113 +762,375 @@ def simulate(
-                    "genome with recombination or gene conversion without "
-                    "rate map"
+                    "When using the DTWF model, the population size must be set "
+                    "explicity, either using the population_size or demography "
```
(minor typo) - explicitly

gtsambos commented on msprime/ancestry.py:
```diff
@@ -721,113 +762,375 @@ def simulate(
-        length=sequence_length,
+    if initial_state is None:
+        if samples is None:
+            raise ValueError("Either samples or initial state must be specified")
```
By initial state you mean a tree sequence inputted with the optional argument `initial_state` here right? I'd make this more specifically clear by using the exact argument name.

jeromekelleher replied:
Done.

gtsambos commented on msprime/ancestry.py:
```diff
@@ -851,6 +1154,9 @@ def __init__(
         ll_tables = _msprime.LightweightTableCollection(tables.sequence_length)
         ll_tables.fromdict(tables.asdict())
 
+        assert len(gene_conversion_map) == 1
```
Is this assertion here because you want the user to specify a uniform gene conversion rate for now? It might be nice to raise a verbose error for the user about this

jeromekelleher replied:
opened #1212 to track.

gtsambos reviewed (approved):
Nice! Only very minor comments from me, though I'm not very familiar with all of this code base so some of it I just scanned. Is there a reason why you changed the various `_parse` methods on Ancestry objects etc. to have the underscore in front? To me that makes them seem like purely internal methods that you aren't expecting people to actively use, but you have invoked them directly in some of the tests -- will users ever need to do this too?

jeromekelleher commented:
> Is there a reason why you changed the various _parse methods on Ancestry objects etc. to have the underscore in front? To me that makes them seem like purely internal methods that you aren't expecting people to actively use, but you have invoked them directly in some of the tests -- will users ever need to do this too?

Excellent question @gtsambos! Putting the underscores in front of the function names is an attempt to clean up the namespace and to make it clear that these *are* internal functions that are not intended for end-users. It's fine for us to use them internally (i.e., for testing), but it's giving a (hopefully) unequivocal signal to end users that they should *not* be using these functions.

jeromekelleher commented:
Thanks very much for the comments all, this was super helpful. I've changed the default ploidy to 2,  implemented other suggestions, and punted a few off to new issues. Let's merge!
